src/util-proto-name.c

```diff
@@ -184,6 +188,265 @@ const char *known_proto[256] = {
+    "mobility-header", /* 0x87: 135 - Header	Mobility Extension Header for IPv6	RFC 6275 */
+    "udplite",         /* 0x88: 136 - Lightweight User Datagram Protocol	RFC 3828 */
+    "mpls-in-ip",      /* 0x89: 137 - Multiprotocol Label Switching Encapsulated in IP	RFC 4023,
+                                RFC      5332 */
+    NULL,              /* 0x8A: 138 - MANET Protocols	RFC 5498 */
+    "hip",             /* 0x8B: 139 - Host Identity Protocol	RFC 5201 */
+    "shim6",           /* 0x8C: 140 - Site Multihoming by IPv6 Intermediation	RFC 5533 */
+    "wesp",            /* 0x8D: 141 - Wrapped Encapsulating Security Payload	RFC 5840 */
+    "rohc",            /* 0x8E: 142 - Robust Header Compression	RFC 5856 */
+    /* no aliases for 142-255 */
+};
+
+typedef struct ProtoNameHashEntry_ {
+    const char *name;
+    const char *alias;
+    uint8_t number;
+} ProtoNameHashEntry;
+
+HashTable *proto_ht;
+
+static uint32_t ProtoNameHashFunc(HashTable *ht, void *data, uint16_t datalen)
+{
+    ProtoNameHashEntry *p = (ProtoNameHashEntry *)data;
+    uint32_t hash = 0;
+    int len = strlen(p->name);
+    for (int i = 0; i < len; i++) {
+        hash += tolower((unsigned char)p->name[i]);
+    }
+
+    return hash % ht->array_size;
+}
+
+static char ProtoNameHashCompareFunc(void *data1, uint16_t datalen1, void *data2, uint16_t datalen2)
+{
+    ProtoNameHashEntry *p1 = (ProtoNameHashEntry *)data1;
+    ProtoNameHashEntry *p2 = (ProtoNameHashEntry *)data2;
+
+    if (p1 == NULL || p2 == NULL)
+        return 0;
+
+    if (p1->name == NULL || p2->name == NULL)
+        return 0;
+
+    int len1 = strlen(p1->name);
+    int len2 = strlen(p2->name);
+
+    return (len1 == len2 &&
+            (memcmp(p1->name, p2->name, len1) == 0 || memcmp(p1->alias, p2->name, len1) == 0));
+}
+
+static void ProtoNameAddEntry(
+        const char *proto_name, const char *alias_name, const int proto_number)
+{
+    ProtoNameHashEntry *proto_ent = SCCalloc(1, sizeof(ProtoNameHashEntry));
+    if (!proto_ent) {
+        FatalError(SC_ERR_HASH_TABLE_INIT, "Unable to allocate protocol hash entry");
+    }
+
+    proto_ent->name = SCStrdup(proto_name);
+    if (!proto_ent->name)
+        goto error_alloc_failed;
+    proto_ent->number = proto_number;
+    if (alias_name) {
+        proto_ent->alias = SCStrdup(alias_name);
+        if (!proto_ent->alias)
+            goto error_alloc_failed;
+    }
+
+    SCLogDebug("new protocol entry: name: \"%s\" [alias: \"%s\"]; protocol number: %d",
+            proto_ent->name, proto_ent->alias ? proto_ent->alias : "n/a", proto_ent->number);
+    if (0 != HashTableAdd(proto_ht, proto_ent, 0)) {
+        FatalError(SC_ERR_HASH_ADD,
+                "Unable to add hash table entry for "
+                "name: \"%s\" [alias: \"%s\"]; number: %d",
+                proto_ent->name, proto_ent->alias, proto_ent->number);
+    }
+    return;
+
+error_alloc_failed:
```

Comment:
I think this use of goto is a bit far fetched... the only reason seems to be to avoid one extra `FatalError` line?

Reply:
There's 2 usages of it but I'll make the update.